rebalance/rebalance.py

```diff
@@ -186,39 +233,60 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi = Non
 
+            time_start = time.time()
+            count_sendpay += 1
             try:
                 plugin.rpc.sendpay(route, payment_hash)
-                plugin.rpc.waitsendpay(payment_hash, retry_for + start_ts - int(time.time()))
-                return success_msg
+                running_for = int(time.time()) - start_ts
+                result = plugin.rpc.waitsendpay(payment_hash, max(retry_for - running_for, 0))
+                time_sendpay += time.time() - time_start
+                if result.get('status') == "complete":
+                    rpc_result["stats"] = f"running_for:{int(time.time()) - start_ts}  count_getroute:{count}  time_getroute:{time_getroute}  time_getroute_avg:{time_getroute / count}  count_sendpay:{count_sendpay}  time_sendpay:{time_sendpay}  time_sendpay_avg:{time_sendpay / count_sendpay}"
+                    return cleanup(plugin, label, payload, rpc_result)
 
             except RpcError as e:
-                plugin.log("RpcError: " + str(e), 'debug')
+                time_sendpay += time.time() - time_start
+                plugin.log(f"maxhops:{maxhops}  msatfactor:{msatfactor}  running_for:{int(time.time()) - start_ts}  count_getroute:{count}  time_getroute:{time_getroute}  time_getroute_avg:{time_getroute / count}  count_sendpay:{count_sendpay}  time_sendpay:{time_sendpay}  time_sendpay_avg:{time_sendpay / count_sendpay}", 'debug')
+                #plugin.log(f"RpcError: {str(e)}", 'debug')
+                # check if we ran into the `rpc.waitsendpay` timeout
+                if e.method == "waitsendpay" and e.error.get('code') == 200:
+                    raise RpcError("rebalance", payload, {'message': 'Timeout reached'})
+                # check if we have problems with our own channels
+                erring_node = e.error.get('data', {}).get('erring_node')
                 erring_channel = e.error.get('data', {}).get('erring_channel')
+                erring_direction = e.error.get('data', {}).get('erring_direction')
                 if erring_channel == incoming_scid:
                     raise RpcError("rebalance", payload, {'message': 'Error with incoming channel'})
                 if erring_channel == outgoing_scid:
                     raise RpcError("rebalance", payload, {'message': 'Error with outgoing channel'})
-                erring_direction = e.error.get('data', {}).get('erring_direction')
+                # exclude other erroring channels
                 if erring_channel is not None and erring_direction is not None:
                     excludes.append(erring_channel + '/' + str(erring_direction))
+                # count and exclude nodes that produce a lot of errors
+                if erring_node:
+                    if nodes.get(erring_node) is None:
+                        nodes[erring_node] = 0
+                    nodes[erring_node] += 1
+                    if nodes[erring_node] > 3:
+                        excludes.append(erring_node)
```

Comment:
About this erring node filter, could we insert this line (or something similar) somewhere here?
```
rpc_result["erring_node_count"] = sum(n > 3 for n in nodes.values())
```
So we may see in the logs, how useful this erring node filter is.